File: adws/trigger_webhook.py
# /// script
# dependencies = ["fastapi", "uvicorn"]
# ///

"""
Webhook trigger for ADW (AI Developer Workflow).

Receives GitHub webhook events and spawns adwPlanBuild.tsx
for new issues. Start with: uv run adws/trigger_webhook.py
"""

import os
import subprocess
import sys
from datetime import datetime, timezone

from fastapi import FastAPI, Header, Request

app = FastAPI()


def log(message: str) -> None:
    timestamp = datetime.now(timezone.utc).isoformat()
    print(f"[{timestamp}] {message}", flush=True)


@app.post("/webhook")
async def webhook(request: Request, x_github_event: str = Header(default="")):
    body = await request.json()

    # Handle PR review comment events
    if x_github_event in ("pull_request_review_comment", "pull_request_review"):
        pr_number = body.get("pull_request", {}).get("number")
        if pr_number is None:
            log("No PR number found in payload")
            return {"status": "ignored"}

        action = body.get("action", "")
        if action not in ("created", "submitted"):
            log(f"Ignored PR review action: {action}")
            return {"status": "ignored"}

        log(f"PR review comment on PR #{pr_number}, triggering ADW PR Review")

        subprocess.Popen(
            ["npx", "tsx", "adws/adwPrReview.tsx", str(pr_number)],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=True,
        )

        return {"status": "triggered", "pr": pr_number}

    if x_github_event != "issues":
        log(f"Ignored event: {x_github_event}")
        return {"status": "ignored"}

    action = body.get("action", "")
    if action != "opened":
        log(f"Ignored issues action: {action}")
        return {"status": "ignored"}

    issue_number = body.get("issue", {}).get("number")
    if issue_number is None:
        log("No issue number found in payload")
        return {"status": "ignored"}

    log(f"New issue #{issue_number} detected, triggering ADW workflow")

    subprocess.Popen(
        ["npx", "tsx", "adws/adwPlanBuild.tsx", str(issue_number)],
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        start_new_session=True,
    )

    return {"status": "triggered", "issue": issue_number}


if __name__ == "__main__":
    import uvicorn

    port = int(os.environ.get("PORT", "8001"))
    log(f"Starting webhook trigger on port {port}")
    uvicorn.run(app, host="0.0.0.0", port=port)

File: adws/trigger_webhook.ts
#!/usr/bin/env npx tsx

/**
 * Webhook trigger for ADW (AI Developer Workflow).
 *
 * Receives GitHub webhook events and spawns adwPlanBuild.tsx
 * for new issues and adwPrReview.tsx for PR review comments.
 * Start with: npx tsx adws/trigger_webhook.ts
 */

import * as http from 'http';
import { spawn } from 'child_process';
import { log } from './utils';

const port = parseInt(process.env.PORT || '8001', 10);

function jsonResponse(
  res: http.ServerResponse,
  statusCode: number,
  body: Record<string, unknown>,
): void {
  res.writeHead(statusCode, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify(body));
}

function spawnDetached(command: string, args: string[]): void {
  const child = spawn(command, args, {
    detached: true,
    stdio: 'ignore',
  });
  child.unref();
}

const server = http.createServer((req, res) => {
  if (req.method !== 'POST' || req.url !== '/webhook') {
    jsonResponse(res, 404, { error: 'not found' });
    return;
  }

  const chunks: Buffer[] = [];
  req.on('data', (chunk: Buffer) => chunks.push(chunk));
  req.on('end', () => {
    let body: Record<string, unknown>;
    try {
      body = JSON.parse(Buffer.concat(chunks).toString());
    } catch {
      jsonResponse(res, 400, { error: 'invalid json' });
      return;
    }

    const event = req.headers['x-github-event'] as string | undefined;

    // Handle PR review comment events
    if (event === 'pull_request_review_comment' || event === 'pull_request_review') {
      const pr = (body.pull_request as Record<string, unknown> | undefined);
      const prNumber = pr?.number as number | undefined;
      if (prNumber == null) {
        log('No PR number found in payload');
        jsonResponse(res, 200, { status: 'ignored' });
        return;
      }

      const action = (body.action as string) || '';
      if (action !== 'created' && action !== 'submitted') {
        log(`Ignored PR review action: ${action}`);
        jsonResponse(res, 200, { status: 'ignored' });
        return;
      }

      log(`PR review comment on PR #${prNumber}, triggering ADW PR Review`);
      spawnDetached('npx', ['tsx', 'adws/adwPrReview.tsx', String(prNumber)]);
      jsonResponse(res, 200, { status: 'triggered', pr: prNumber });
      return;
    }

    // Handle issue events
    if (event !== 'issues') {
      log(`Ignored event: ${event || '(none)'}`);
      jsonResponse(res, 200, { status: 'ignored' });
      return;
    }

    const action = (body.action as string) || '';
    if (action !== 'opened') {
      log(`Ignored issues action: ${action}`);
      jsonResponse(res, 200, { status: 'ignored' });
      return;
    }

    const issue = (body.issue as Record<string, unknown> | undefined);
    const issueNumber = issue?.number as number | undefined;
    if (issueNumber == null) {
      log('No issue number found in payload');
      jsonResponse(res, 200, { status: 'ignored' });
      return;
    }

    log(`New issue #${issueNumber} detected, triggering ADW workflow`);
    spawnDetached('npx', ['tsx', 'adws/adwPlanBuild.tsx', String(issueNumber)]);
    jsonResponse(res, 200, { status: 'triggered', issue: issueNumber });
  });
});

log(`Starting webhook trigger on port ${port}`);
server.listen(port, '0.0.0.0', () => {
  log(`Webhook server listening on 0.0.0.0:${port}`);
});
